### ISSUE TITLE
Feature Request: Implement a touchbar API

### ISSUE THREAD

**Issue by appetrosyan** [aquamacs-emacs/aquamacs-emacs#135]
Would be nice if there was a way of creating User-customiseable touchbar buttons and widgets. It's still possible through better Touch tool, but this approach has limitations. 
1) The buttons have significant overhead. In order to call an emacs function, one needs to send option-x the command and then an enter event, which is sub optimal. 
2) Buttons can't be natively context sensitive. There's a way of making buttons become invisible depending on the return of an AppleScript, but that's both inefficient and unreliable. 
3) Buttons can't be programmatically added. I.e. one needs to update the better touch tool configuration in order to make the new functions available. 

If the y/n prompts could be redirected to the touchbar, that'd be great.

**Comment by davidswelt** [comment]
This should be quite doable.  You would need a C-level API that exposes some functions like it was done for the toolbar -- mostly via keymaps.  This would probably be tied to the selected window or the buffer shown in the selected window.  The toolbar is an unholy mix of keymaps, as far as I know...

**Comment by maehne** [comment]
Native support for the touchbar in Aquamacs would be great to quickly access context-dependent editing functions. How about mirroring/moving the toolbar buttons to the touchbar by default? The toolbar content is already mode-dependent, but costs precious vertical screen space.

**Comment by appetrosyan** [comment]
I'd be up for an API. If there were a way of using keymaps with the touchbar, I'm sure there'd be many spins on what to use it for. I'd use it to mirror some mode specific menus, because pressing a virtual button is way faster than doing e.g. `C-c M-[ C-S-;`, not to mention you don't have to memorise as many mode specific keybinds.

**Comment by davidswelt** [comment]
That can be done.  New images would have to be designed (or at least experimented with).

The question is - who has time and expertise to implement it?
I’d gladly give feedback for an implementation, and I’d suggest to discuss a concept first./

David

**Comment by Digicrat** [comment]
Any progress here?

In lieu of this, I was thinking of a generic approach using BetterTouchTool to create major-mode specific buttons & related widgets.  My thought is to use "emacsclient -e" to query information from emacs/Aquamacs in a bash script that can provide input to BTT (or other tools in the future) for button labels.  Actions could then either be sent as keypresses via AppleScript (easy, but inelegant) or again via emacsclient (which may require more LISP magic than I have).  

My question though is how do I get that information from emacsclient/Aquamacs?  From there, scripting dynamic buttons with BTT is fairly straight forward (even if not very pretty).

It looks like (selected-frame) should give me a reference to the currently active frame, but how do I query from there the: (a) buffer name, (b) major-mode, (c) minor-modes, (d) line number, etc. ?  (Note my LISP is just a bit rusty).

**Comment by davidswelt** [comment]
Yeah, a generic interface through emacsclient would be great, although implementing TouchBar support this way seems rather hackt and brittle.  Consider latency.

All of the items you mention are relatively easy to get, and documentation should be readily available.  For variables such as major-mode, you can wrap them in (with-current-buffer ...).

**Comment by Digicrat** [comment]
Admittedly the latency isn't great, but a 1-second delay isn't terrible for just triggering on the active buffer/major-mode.

I can't quite seem to get the information I want though, as the emacsclient executes from the context of its own frame.

(selected-frame)  - Gives me a reference to the desired active frame (and shows a not-pretty name of it if printed)

(with-current-buffer) should let me get information (ie: major mode) from a given buffer, but it takes a reference to a buffer as input.  I can't figure out how to get the active buffer from the selected-frame.  

`emacsclient -e '(with-selected-frame (selected-frame) (buffer-name))'`
Theoretically with-selected-frame should be what I want (it exists, though I can't find any documentation on it), but it always returns '*server*'.

Any ideas?  Thanks.

**Comment by Digicrat** [comment]
As I was posting the above question (it always bugs me when something simple doesn't work ;-), I came across information on BTT's Web Server.  http://docs.bettertouchtool.net/docs/1104_webserver.html .  

This would still be a messy solution, but it should be possible to write a hook in emacs that pushes data to BTT as appropriate via the REST API to create something a bit more responsive and hopefully less fragile.  Specifically, the update_touch_bar_widget can be used to update text+icons, but not the action.  The action could either be a generic keybinding that we bind appropriately within said script, or use emacsclient to call a custom function with the key number as a parameter (which would have the same issues as above in identifying the active buffer).

**Comment by davidswelt** [comment]
Why don’t you implement a lisp interface in C so that one can set it on the Lisp side (via a keymap)?
This seems 1000 times more straightforward, and it _will_ be reliable.

**Comment by davidswelt** [comment]
Get the active window and then the buffer for that.

**Comment by appetrosyan** [comment]
Any progress? There doesn't seem to be any chance that apple falls back from making the Touch Bar mandatory. Sadly, my favourite text editor doesn't let me use that.

**Comment by HayoBaan** [comment]
I guess creating a customized version of the Touch Bar is probably not straightforward (to integrate in Aquamacs) so this is likely not going to happen (soon). Instead why don't you configure Aquamacs as an application that uses standard function keys?
Go to System Preferences | Keyboard | Shortcuts, select "fn Function Keys" on the left side of the panel, click the + button and select Aquamacs. Et Voila you have standard function keys in Aquamacs 😃

**Comment by rpgoldman** [comment]
Is there any way to get a Super modifier onto the touch bar?  I'd love to have a dedicated Super key, but there isn't really a place to put it on an ordinary keyboard...

**Comment by HayoBaan** [comment]
No there isn't, but you could do what I have done and assign one of the CMD keys as super modifier key. You can customize this using Options|Customize Aquamacs|Top Level Customization Group|Aquamacs and then customize e.g. the `Ns Command Modifier` or `Ns Right Command Modifier`. I choose the right command modifier to act as “super” (and the right alternate modifier as “none” so I could always use it to enter e.g. ümlauts and other special characters).

**Comment by Digicrat** [comment]
If you use BetterTouchTool (BTT), you can effectively put anything that you want in the touch bar.  

I've finally gotten back around to looking at this again and, after some
trials & tribbleations, I've come to a solution that may live long and
prosper (for me) -- as well as significantly improving my Emacs-Lisp-fu.

I've written a series of functions and keybindings in emacs that utilize BTTs HTTP server for sending updates.  BTT in turn simply sends keystrokes to interact with Emacs. I tried other approaches, including emacsclient, but this proved to be the most responsive.  

As a bonus, this approach is generic and should be usable with any version of emacs with minimal effort.  If you enable remote access, it should even work from within a VM (given appropriate BTT setup).

My touchbar for emacs currently displays the Major Mode, File Name and Line number.  You can click the file name for a list of the 10 most-recently viewed buffers to jump to.  I also have widgets for bm.el and eyebrowse (two plugins I use often), and intend on adding more major-mode-specific options as I think/need them later on.

There's much more that can be done (or cleaned up), but this is a good start (for my needs at least).

See https://github.com/Digicrat/btt for details.
